### ISSUE TITLE
`COPY --from` breaks `docker build --target`

### ISSUE THREAD

**Issue by PlushBeaver** [moby/moby#37056]
**Description**

I'm trying to split build environment container into a stage with `programs` (for CI) and a later stage with `tools` (for developers). However, having a `COPY --from` instruction in the `programs` stage makes it impossible to build this stage only. Removing `COPY --from` resolves the issue.

**Steps to reproduce the issue:**

1. Create a Dockerfile that causes the error:

    ```
    FROM alpine AS programs
    RUN apk update && apk add less
    COPY --from=alpine /bin/busybox /test

    FROM programs AS tools
    COPY script /opt/script
    ```

2. Try to build `programs` stage:

    ```
    docker build --target programs .
    ```

**Describe the results you received:**

Build does not stop at `programs` stage and finished with an error:

```
Sending build context to Docker daemon   2.56kB
Step 1/5 : FROM alpine AS programs
 ---> 3fd9065eaf02
Step 2/5 : RUN apk update && apk add less
 ---> Running in 02c6480fb552
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/x86_64/APKINDEX.tar.gz
v3.7.0-162-g06efc38960 [http://dl-cdn.alpinelinux.org/alpine/v3.7/main]
v3.7.0-161-gd98d33b7f4 [http://dl-cdn.alpinelinux.org/alpine/v3.7/community]
OK: 9050 distinct packages available
(1/4) Installing ncurses-terminfo-base (6.0_p20171125-r0)
(2/4) Installing ncurses-terminfo (6.0_p20171125-r0)
(3/4) Installing ncurses-libs (6.0_p20171125-r0)
(4/4) Installing less (520-r0)
Executing busybox-1.27.2-r7.trigger
OK: 12 MiB in 15 packages
 ---> 2c131388defe
Removing intermediate container 02c6480fb552
Step 3/5 : COPY --from=alpine /bin/busybox /test
 ---> b0ab64389695
Removing intermediate container a46427467d3d
Step 4/5 : FROM programs AS scripts
 ---> b0ab64389695
Step 5/5 : COPY script /opt/script
 ---> 3f5c853dfddb
Removing intermediate container 31eb07681ff2
failed to reach build target programs in Dockerfile
```

**Describe the results you expected:**

Build stops after `COPY --from` instruction.

**Output of `docker version`:**

```
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:09:06 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:09:06 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 6
 Running: 0
 Paused: 0
 Stopped: 6
Images: 1261
Server Version: 17.05.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.4.6
Operating System: Debian GNU/Linux buster/sid
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.56GiB
Name: kozlyuk-pc
ID: SQTS:FBMC:65JL:X4RO:HF76:YFI2:JFQD:SCUK:AJ3R:WTPN:QR52:RDPS
Docker Root Dir: /home/kozlyuk/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://kozlyuk:4iDCf8pu5j@proxy.bifit.int:3128
Https Proxy: http://kozlyuk:4iDCf8pu5j@proxy.bifit.int:3128
No Proxy: docker.netdike,mirrors.netdike,sv1.netdike
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
```

**Comment by AkihiroSuda** [comment]
Could you try recent version?

**Comment by PlushBeaver** [comment]
Issue appears fixed in `18.03.1~ce-0~debian`. Thank you and sorry for the trouble.
